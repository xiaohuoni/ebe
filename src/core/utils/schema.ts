import { Component } from '@lingxiteam/types';
import * as _ from 'lodash';
import { IPublicTypeNodeDataType } from '../types';
import { CodeGeneratorError } from '../types/error';
import {
  isDOMText,
  isJSExpression,
  isJSFunction,
  isJSSlot,
  isNodeSchema,
} from '../utils/deprecated';
import { isJSExpressionFn } from './common';
export function isContainerSchema(x: any): x is any {
  return (
    typeof x === 'object' &&
    x &&
    typeof x.type === 'string' &&
    typeof x.fileName === 'string'
  );
}

export function isNpmInfo(x: any): x is any {
  return typeof x === 'object' && x && typeof x.package === 'string';
}

const noop = () => undefined;

const handleChildrenDefaultOptions = {
  rerun: false,
};

const DEFAULT_MAX_DEPTH = 100000;

/**
 * 遍历并处理所有的子节点
 * @param components
 * @param handlers
 * @param options
 * @returns
 */
export function handleSubNodes<T>(
  components: IPublicTypeNodeDataType,
  handlers: {
    string?: (i: string) => T;
    expression?: (i: any) => T;
    node?: (i: IPublicTypeNodeDataType) => T;
  },
  options?: {
    rerun?: boolean;
    maxDepth?: number; // 防止出现死循环无穷递归
  },
): T[] {
  const opt = {
    ...handleChildrenDefaultOptions,
    ...(options || {}),
  };
  const maxDepth = opt.maxDepth ?? DEFAULT_MAX_DEPTH;
  if (maxDepth <= 0) {
    throw new Error('handleSubNodes maxDepth reached');
  }

  if (Array.isArray(components)) {
    const list: any[] = components as any;
    return list
      .map((child) =>
        handleSubNodes(child, handlers, { ...opt, maxDepth: maxDepth - 1 }),
      )
      .reduce((p, c) => p.concat(c), []);
  }

  let result: T | undefined;
  const childrenRes: T[] = [];
  if (components === null || components === undefined) {
    return [];
  } else if (isDOMText(components)) {
    const handler = handlers.string || noop;
    result = handler(components as any);
  } else if (isJSExpression(components)) {
    const handler = handlers.expression || noop;
    result = handler(components);
  } else if (isJSSlot(components)) {
    // @ts-ignore
    return handleSubNodes(components.value as any, handlers, {
      ...opt,
      maxDepth: maxDepth - 1,
    });
  } else if (isNodeSchema(components)) {
    const handler = handlers.node || noop;
    const child = components as any;
    result = handler(child);

    if (child.components) {
      const childRes = handleSubNodes(child.components, handlers, opt);
      childrenRes.push(...childRes);
    }

    if (child.props) {
      if (Array.isArray(child.props)) {
        child.props.forEach(({ value }: any) => {
          const childRes = handleCompositeValueInProps(value);
          childrenRes.push(...childRes);
        });
      } else {
        Object.values(child.props).forEach((value) => {
          const childRes = handleCompositeValueInProps(value as any);
          childrenRes.push(...childRes);
        });
      }
    }
  } else {
    throw new CodeGeneratorError(
      'handleSubNodes got invalid NodeData',
      components,
    );
  }

  if (result !== undefined) {
    childrenRes.unshift(result);
  }

  return childrenRes;

  function handleCompositeValueInProps(value: any): T[] {
    if (isJSSlot(value)) {
      return handleSubNodes(value.value, handlers, {
        ...opt,
        maxDepth: maxDepth - 1,
      });
    }

    // CompositeArray
    if (Array.isArray(value)) {
      return _.flatMap(value, (v) => handleCompositeValueInProps(v));
    }

    // IPublicTypeCompositeObject
    if (
      !isJSExpression(value) &&
      !isJSExpressionFn(value) &&
      !isJSFunction(value) &&
      typeof value === 'object' &&
      value !== null
    ) {
      return _.flatMap(Object.values(value), (v) =>
        handleCompositeValueInProps(v),
      );
    }

    return [];
  }
}

export function isValidContainerType(schema: Component) {
  return ['Page', 'Component', 'Block'].includes(schema.type);
}

export const enum ContainerType {
  Page = 'Page',
  Component = 'Component',
  Block = 'Block',
}
