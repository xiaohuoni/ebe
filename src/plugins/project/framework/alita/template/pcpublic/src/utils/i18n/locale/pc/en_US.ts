import merge from 'lodash/merge';
import { baseEnUS } from '../base';

export default merge(baseEnUS, {
  'Quill.uploadImages':
    'Only images in the formats of .jpg, .gif, and .png can be uploaded.',
  'Quill.imageSizeLimit': 'Each image cannot be larger than 3000KB',
  'Quill.selectZh': 'Please select the text before clicking on the link.',
  'Table.pageNotEmptyWarning': 'current page number is empty.',
  'Tree.appendTreeNodeWarning':
    'The data of the child nodes of the loaded tree must have the key value, without the key can not be manipulated.',
  'Tree.appendMenuNodeWarning':
    'Load menu node must have key value, without key can not be manipulated.',
  'Tree.editingWarn': 'There are no tree nodes being manipulated.',
  'Audio.cannotPlay': 'Audio node removed, can not be played.',
  'Audio.cannotStop': 'Audio node removed, can not be paused.',
  'Audio.cannotSwitch': 'Audio node removed, can not be switched.',
  'Audio.warning': 'Error getting audio file.',
  'Icon.imgFormatWarning': 'Image format error',
  'Input.lenRule':
    'Content length must be greater than or equal to @{minLength} and less than or equal to @{maxLength}.',
  'Input.maxRule':
    'The length of the content must be less than or equal to @{maxLength}.',
  'Input.minRule':
    'Content length must be greater than or equal to @{minLength}',
  'Upload.download': 'The file is downloading, please be patient.',
  'Upload.readonly':
    'Currently read-only, no files are allowed to be uploaded or removed!',
  'Upload.maxSize':
    'The current maximum file upload size is @{singleFileMaxSize}, please try again!',
  'Upload.minSize':
    'The current minimum file upload size is @{singleFileMinSize}, please try again!',
  'Upload.maxNum':
    'The current maximum number of files to upload is @{numberLimit}, no further uploads are allowed.',
  'Upload.format':
    'The file format is not correct, please upload the file in @{acceptType} format.',
  'Upload.limit':
    'The current maximum number of files to upload is @{numberLimit}, please try again.',
  'Upload.text': 'Click or drag the file into the dotted box to upload it.',
  'Upload.uploadText': 'Click on Upload',
  'Upload.downloadText': 'download',
  'Upload.origin': 'Download source file',
  'Upload.waterMark': 'Download file with watermark',
  'Upload.downloadType': 'file download options',
  'Pagination.showTotal': 'Total @{tol} items',
  'RadarChart.warning': 'Dimension data is incorrectly configured.',
  'RemoteComponent.fail': 'Failed to load custom component.',
  'Steps.wait': 'Waiting',
  'Steps.process': 'In Process',
  'Steps.finish': 'Finished',
  'Steps.error': 'Error',
  'Table.dateTypeWarning': 'Inconsistent data types',
  'Table.headTips':
    'Columns can be reordered by dragging and dropping the icons on the left.',
  'Table.headCol': 'columns',
  'Table.needObj':
    'Add rows of data that should be formatted as objects, but receive non-object values.',
  'VerificationCode.text': 'Send verification code',
  'VerificationCode.refresh': 'Refresh',
  'Img.fileIdError':
    'Command error: received value does not match, current value is :@{field}, expects a number or string.',
  'Form.actionWarning':
    'This component does not support the use of the getFormValues directive.',
  'Transfer.unit': '@{num} items',
  'Transfer.min': 'minimum @{min} data for @{name} items.',
  'Transfer.max': 'Maximum @{max} data for @{name} items.',
  'Table.filter.empty': 'empty', //
  'Table.expand.component': 'Business components have been associated:',
  'Table.data.typeError':
    'Form data type error, current type is @{type}, please check configuration.',
  'SuperSelect.load': 'Only 50 data loaded, for more data please search',
  'DatePicker.afterToday': '@{name} must be after (and including) today',
  'DatePicker.beforeToday': '@{name} must be before (and including) today',
  'DatePicker.period':
    '@{name} must be in the range @{startTime} to @{endTime}',
  'DatePicker.after': 'The @{name} time cannot be smaller than @{startTime}.',
  'DatePicker.before': 'The @{name} time cannot be greater than the @{endTime}',
  'ExpBusiObjModal.title': 'Export Business Object Data by Field',
  'ExpBusiObjModal.exportFieldError':
    'Please select the fields to export first',
  'ExpSQLServiceModal.title': 'Export SQL Query Service Data by Field',
  'ExpSQLServiceModal.importData': 'Import Data',
  'ExpSQLServiceModal.importFieldError':
    'Please select the fields to import first',
  'ExpBusiObjModal.downloadMutiTpl': 'Download Batch Import Template',
  'ExpBusiObjModal.downloadTpl': 'Download Import Template',
  'ExpBusiObjModal.dataDealing': 'Data processing...',
  'ExpBusiObjModal.downloadFailReason':
    'Download failed import data, check the reason for failure',
  'ExpBusiObjModal.validField': 'Specify valid fields for import',
  'ExpBusiObjModal.dragView':
    'Click to select a file or drag a file to this area for upload',
  'ExpBusiObjModal.onlySingleUpload': 'Support single file upload only',
  'ExpBusiObjModal.validTable': 'Please upload a valid table',
  'ExpBusiObjModal.uploadFail': '@{name} file upload failed, please try again!',
  'ExpBusiObjModal.noFile': 'No file uploaded, please select a file first',
  'ExpBusiObjModal.importFailList': 'Import failed list-@{name}.xlsx',
  'ExpBusiObjModal.partImportFailTitle': 'Partial data import failed',
  'ExpBusiObjModal.importSuccessTitle': 'Import successful',
  'ExpBusiObjModal.importSuccessMsg':
    'Data import was successful, please choose to go to the list to view',
  'ExpBusiObjModal.importFailTitle': 'Import failed',
  'ExpBusiObjModal.importFailMsg':
    'Data import failed, please check the following information and resubmit.',
  'AccountDivideModal.setAccount': 'Set Account',
  'AccountDivideModal.edit': 'Change Plan',
  'AccountDivideModal.detail': 'Plan Details',
  'AccountDivideModal.disableAutoFill':
    'Does not meet the automatic fill condition',
  'AccountDivideModal.fillFail':
    'Automatic filling failed, please manually enter at least one basis data and try again',
  'AccountDivideModal.AccountNoFinish':
    'Account division not completed, please check carefully',
  'AccountDivideModal.validDate': 'Please select a valid start date',
  'AccountDivideModal.startDatePlaceholder': 'Please select a valid start date',
  'AccountDivideModal.accountMode': 'Billing Mode',
  'ProgressComp.cancelMessage':
    'Canceling will result in no download results. Please confirm if you want to cancel',
  'ProgressComp.progressMsg': '@{fileName} is in progress: @{text}',
  'ImportBusiObjModal.tips':
    'Import total records @{total}, of which @{success} are successful, and @{fail} are failures',
  'AccountDivideDetail.subject': 'Subject:',
  'AccountDivideDetail.plan': 'Current Plan',
  'AccountDivideDetail.aleterAfter': 'Plan After Change',
  'Upload.uploadPlaceholder': 'Click or drag and drop files here to upload',
  'Upload.singleStrict':
    'Single file mode does not allow uploading multiple files',
  'Upload.actionStrict':
    'The upload action is not configured, please return to modify it!',
  'Upload.fileLimit': 'Exceeds file size limit, currently @{num}',
});
