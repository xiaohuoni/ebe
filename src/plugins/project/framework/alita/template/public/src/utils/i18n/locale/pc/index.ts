import zhCN from './zh_CN';

export default {
  zhCN: () => Promise.resolve(zhCN),
  other: () =>
    import(/* webpackChunkName: 'other-pc-locale' */ './otherLocale'),
};
