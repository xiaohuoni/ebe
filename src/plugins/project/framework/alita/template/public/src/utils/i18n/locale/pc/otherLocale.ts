export { default as enUS } from './en_US';
export { default as zhHK } from './zh_HK';
