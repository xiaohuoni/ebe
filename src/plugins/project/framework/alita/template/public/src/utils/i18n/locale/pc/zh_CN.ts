import merge from 'lodash/merge';
import { baseZHCN } from '../base';

export default merge(baseZHCN, {
  'Quill.uploadImages': '只能上传格式为.jpg,.gif,.png的图片',
  'Quill.imageSizeLimit': '每张图片不能大于3000KB',
  'Quill.selectZh': '请先选中文字，再点击链接',
  'Table.pageNotEmptyWarning': '当前页码为空',
  'Tree.appendTreeNodeWarning': '加载树的子节点数据必须有key键值，没有key不能操作',
  'Tree.appendMenuNodeWarning': '加载菜单节点必须有key键值，没有key不能操作',
  'Tree.editingWarn': '没有正在操作的树节点',
  'Audio.cannotPlay': '音频节点被移除,无法播放',
  'Audio.cannotStop': '音频节点被移除,无法暂停',
  'Audio.cannotSwitch': '音频节点被移除,无法切换',
  'Audio.warning': '音频文件获取出错',
  'Icon.imgFormatWarning': '图片格式错误',
  'Input.lenRule': '内容长度须大于等于@{minLength}且小于等于@{maxLength}',
  'Input.maxRule': '内容长度须小于等于@{maxLength}',
  'Input.minRule': '内容长度须大于等于@{minLength}',
  'Upload.download': '该文件正在下载，请耐心等待',
  'Upload.readonly': '当前为只读状态，不允许上传或移除文件!',
  'Upload.maxSize': '当前文件最大上传大小为@{singleFileMaxSize}，请重新上传',
  'Upload.minSize': '当前文件最小上传大小为@{singleFileMinSize}，请重新上传',
  'Upload.maxNum': '当前文件最大上传个数为@{numberLimit}，不允许继续上传',
  'Upload.format': '文件格式不正确，请上传@{acceptType}格式的文件',
  'Upload.limit': '当前文件最大上传个数为@{numberLimit}，请重新选择',
  'Upload.text': '点击或者拖动文件到虚线框内上传',
  'Upload.uploadText': '点击上传',
  'Upload.downloadText': '下载',
  'Upload.origin': '源文件下载',
  'Upload.waterMark': '加水印下载',
  'Upload.downloadType': '文件下载方式',
  'Pagination.showTotal': '共 @{tol} 条',
  'RadarChart.warning': '维度数据配置有误',
  'RemoteComponent.fail': '自定义组件加载失败',
  'Steps.wait': '等待状态',
  'Steps.process': '处理中状态',
  'Steps.finish': '完成状态',
  'Steps.error': '错误状态',
  'Table.dateTypeWarning': '数据类型不一致',
  'Table.headTips': '通过左侧图标拖拽可调整列顺序',
  'Table.headCol': '列设置',
  'Table.needObj': '添加行数据的格式应该为对象，但接受到非对象的值',
  'VerificationCode.text': '发送验证码',
  'VerificationCode.refresh': '重新获取',
  'Img.fileIdError': '指令错误：接收的值不匹配，当前值为:@{field}, 期望为数字或字符串',
  'Form.actionWarning': '该组件不支持使用getFormValues指令',
  'Transfer.unit': '@{num}项',
  'Transfer.min': '@{name}最少需要选@{min}条数据',
  'Transfer.max': '@{name}最多可选@{max}条数据',
  'Table.filter.empty': '空白',
  'Table.expand.component': '已关联业务组件：',
  'Table.data.typeError': '表格数据类型错误，当前类型为@{type}，请检查配置',
  'SuperSelect.load': '仅加载50条数据，更多数据请搜索',
  'DatePicker.afterToday': '@{name}必须在今天之后（含今天）',
  'DatePicker.beforeToday': '@{name}必须在今天之前（含今天）',
  'DatePicker.period': '@{name}必须在@{startTime}到@{endTime}的范围内',
  'DatePicker.after': '@{name}不能小于@{startTime}',
  'DatePicker.before': '@{name}不能大于@{endTime}',
  'ExpBusiObjModal.title': '按字段导出业务对象数据',
  'ExpBusiObjModal.exportFieldError': '请先选择要导出的字段',
  'ExpSQLServiceModal.title': '按字段导出SQL查询服务数据',
  'ExpSQLServiceModal.importData': '导入数据',
  'ExpSQLServiceModal.importFieldError': '请先选择要导入的字段',
  'ExpBusiObjModal.downloadMutiTpl': '下载批量导入模版',
  'ExpBusiObjModal.downloadTpl': '下载导入模版',
  'ExpBusiObjModal.dataDealing': '数据处理中...',
  'ExpBusiObjModal.downloadFailReason': '下载导入失败数据，查看失败原因',
  'ExpBusiObjModal.validField': '指定导入的有效字段',
  'ExpBusiObjModal.dragView': '点击选择文件或拖动文件到此区域进行上传',
  'ExpBusiObjModal.onlySingleUpload': '仅支持单个文件上传',
  'ExpBusiObjModal.validTable': '请上传合法的表格',
  'ExpBusiObjModal.uploadFail': '@{name}文件上传失败，请重试！',
  'ExpBusiObjModal.noFile': '无上传文件，请先选择文件',
  'ExpBusiObjModal.importFailList': '导入失败列表-@{name}.xlsx',
  'ExpBusiObjModal.partImportFailTitle': '部分数据导入失败',
  'ExpBusiObjModal.importSuccessTitle': '导入成功',
  'ExpBusiObjModal.importSuccessMsg': '数据导入成功，请选择前往列表查看',
  'ExpBusiObjModal.importFailTitle': '导入失败',
  'ExpBusiObjModal.importFailMsg': '数据导入失败，请核对以下信息后，再重新提交。',
  'AccountDivideModal.setAccount': '设置分账',
  'AccountDivideModal.edit': '变更计划',
  'AccountDivideModal.detail': '计划详情',
  'AccountDivideModal.disableAutoFill': '不符合自动填充条件',
  'AccountDivideModal.fillFail': '自动填充失败，请至少先手动填入一项依据数据然后重试',
  'AccountDivideModal.AccountNoFinish': '分账未完成，请仔细检查',
  'AccountDivideModal.validDate': '请选择有效的开始日期',
  'AccountDivideModal.startDatePlaceholder': '请选择有效的开始日期',
  'AccountDivideModal.accountMode': '出账方式',
  'ProgressComp.cancelMessage': '取消将无法获得下载结果，请确认是否需要取消',
  'ProgressComp.progressMsg': '@{fileName} 正在@{text}：',
  'ImportBusiObjModal.tips': '导入总记录数@{total}条，其中@{success}条成功，@{fail}条失败',
  'AccountDivideDetail.subject': '科目：',
  'AccountDivideDetail.plan': '当前计划',
  'AccountDivideDetail.aleterAfter': '变更后计划',
  'Upload.uploadPlaceholder': '点击或拖拽文件到此处上传',
  'Upload.singleStrict': '单文件模式不允许上传多个文件',
  'Upload.actionStrict': '上传接口未配置，请返回修改！',
  'Upload.fileLimit': '超出文件数量限制，当前为@{num}',
});
