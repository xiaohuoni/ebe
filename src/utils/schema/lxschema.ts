import { PlatformType, ProcessFunctionType } from '@lingxiteam/types';
import * as _ from 'lodash';
import { LOOPCOMPONENTS, LoopMarkSymbol } from '../../constants';
import { IProjectSchema, IPublicTypeNodeDataType } from '../../core';
import { CodeGeneratorError } from '../../core/types/error';
import { generateVarString, parse2Var } from '../../core/utils/compositeType';
import { isJSVar, isNodeSchema } from '../../core/utils/deprecated';
import assetHelper from './assets/assets';
import { parseDsl } from './parseDsl';

const noop = () => undefined;

const handleChildrenDefaultOptions = {
  rerun: false,
};

/**
 * 生成uid
 * @param schema
 * @returns
 */
export const generateUid = (schema: IProjectSchema) => {
  let uid = parse2Var(schema.id);
  // 如果在循环容器下
  if (schema[LoopMarkSymbol]?.parentLoopId) {
    uid = `getUid(props.compId, props.itemId, ${parse2Var(schema.id)})`;
  }
  return uid;
};

const DEFAULT_MAX_DEPTH = 100000;

/**
 * 遍历并处理所有的子节点
 * @param components
 * @param handlers
 * @param options
 * @returns
 */
export function handleSubNodes<T>(
  components: IPublicTypeNodeDataType,
  handlers: {
    string?: (i: string) => T;
    node?: (i: IPublicTypeNodeDataType) => T;
  },
  options?: {
    rerun?: boolean;
    maxDepth?: number; // 防止出现死循环无穷递归
  },
): T[] {
  const opt = {
    ...handleChildrenDefaultOptions,
    ...(options || {}),
  };
  const maxDepth = opt.maxDepth ?? DEFAULT_MAX_DEPTH;
  if (maxDepth <= 0) {
    throw new Error('handleSubNodes maxDepth reached');
  }

  if (Array.isArray(components)) {
    const list: IPublicTypeNodeDataType = components;
    return list
      .map((child) =>
        handleSubNodes(child, handlers, { ...opt, maxDepth: maxDepth - 1 }),
      )
      .reduce((p, c) => p.concat(c), []);
  }

  let result: T | undefined;
  const childrenRes: T[] = [];
  if (components === null || components === undefined) {
    return [];
  } else if (isNodeSchema(components)) {
    const handler = handlers.node || noop;
    const child = components;
    result = handler(child);

    if (child.components) {
      const childRes = handleSubNodes(child.components, handlers, opt);
      childrenRes.push(...childRes);
    }
  } else {
    throw new CodeGeneratorError(
      'handleSubNodes got invalid NodeData',
      components,
    );
  }

  if (result !== undefined) {
    childrenRes.unshift(result);
  }

  return childrenRes;
}

// TODO: extraData sandBoxContext
const StaticDataKey = 'attrDataMap';

export const extraData = {
  collectRelationKey: (
    compId: string,
    attrNbrKey: string,
    childKeys?: string[],
  ) => {
    if (!attrNbrKey) {
      return '';
    }
    if (Array.isArray(childKeys) && childKeys.length > 0) {
      return `$${StaticDataKey}['${attrNbrKey}'].filter(item=>${JSON.stringify(
        childKeys,
      )}.includes(item.attrValue))$`;
    }
    return `$${StaticDataKey}['${attrNbrKey}']$`;
  },
} as any;
/**
 * 控件属性预处理-只执行一次
 * @param schema
 * @param isRoot
 * @returns
 */
const preprocessComponentSchema = (
  schema: IProjectSchema,
  isRoot: boolean,
): IProjectSchema => {
  let newSchema: any = _.cloneDeep(schema);

  const sandBoxContext = {};
  // 预处理事件
  newSchema = parseDsl(schema, isRoot);

  if (isRoot) {
    return newSchema;
  }

  // 业务组件和页面容器不需要预处理
  if (schema.compName && ['Pageview', 'BOFramer'].includes(schema.compName)) {
    if (schema.compName === 'BOFramer') {
      schema.props.state = schema.props.busiCompStates;
      delete schema.props.busiCompStates;
    } else if (schema.compName === 'Pageview') {
      schema.props.state = schema.props.pageViewCompState;
      delete schema.props.pageViewCompState;
    }

    // newSchema.props = {
    //   ...newSchema.props,
    //   $$componentItem: `##{{id: '${schema.id}',uid: ${generateUid(
    //     schema,
    //   )}, type: '${schema.compName}'}}##`,
    // };
    // return newSchema;
  }
  const methods = assetHelper.comPreprocess.getComPreprocessMethods(
    schema.compName!,
    schema.platform as PlatformType,
  );

  methods?.forEach((fc: ProcessFunctionType) => {
    newSchema = fc(newSchema, extraData);
  });

  const originProps = schema?.props || {};

  const props = {
    ...originProps,
    // pageId: pageId,
    // appId: appId,
    // platform: schema.platform,
    // TODO: fusionMode
    // fusionMode: schema?.fusionMode,
    // uid: `$${generateUid(schema)}$`,
    $$componentItem: `##{{id: '${schema.id}',uid: ${generateUid(schema)}}}##`,
  };

  let className = `${schema.compName}_${schema.id}`;
  if (props.className) {
    // className = `$\`${className} \${${parse2Var(props.className)})}\`$`
    className = `$classNames(${parse2Var(className)}, ${parse2Var(
      props.className,
    )})$`;
  }

  props.className = className;

  // 执行组件预处理
  const methodsRun = assetHelper.comRunPreprocess.getRunComPreprocessMethods(
    schema.compName!,
    schema.platform as PlatformType,
  );

  methodsRun?.forEach((fc: ProcessFunctionType) => {
    // @ts-ignore
    const aprops = fc(newSchema, props, sandBoxContext, extraData);
    newSchema.props = aprops;
  });

  // TOTO: popoverSetting pc
  const { popoverSetting, basicStatus, collapseColumns, ...otherProps } =
    newSchema?.props ?? {};
  // {value: 'popoverSetting={{"options": {"tipType": "2",\n…"destroyOnHide": false,\n"placement": "top"}}}', type: 'NodeCodePieceAttr'}
  if (popoverSetting) {
    // 如果是弹窗的话，需要用组件包裹原始组件
    // @ts-ignore
    return {
      compLib: 'antd-mobile-5',
      type: 'Popover',
      events: {},
      props: { ...popoverSetting, id: otherProps?.['data-compid'] },
      compName: 'Popover',
      platform: 'h5',
      components: [{ ...newSchema, props: otherProps }],
    };
  }
  if (isJSVar(basicStatus)) {
    const basicStatusStr = generateVarString(basicStatus);
    newSchema.props.disabled = `$\`\${${basicStatusStr}}\` === '3'$`;
    newSchema.props.visible = `$\`\${${basicStatusStr}}\` !== '2'$`;
    newSchema.props.readOnly = `$\`\${${basicStatusStr}}\` === '4'$`;
  }

  if (newSchema?.props?.basicStatus) {
    delete newSchema.props.basicStatus;
  }
  // 预处理中已经将 collapseColumns 转成子集，这里删掉
  if (collapseColumns) {
    delete newSchema.props.collapseColumns;
  }
  return newSchema;
};

/**
 * 标记循环容器组件
 * @param schema
 */
export const markerLoopComponent = (schema: IProjectSchema) => {
  const marker = (
    components: IProjectSchema[],
    parentLoopId = '',
    type: any = '',
  ): IProjectSchema[] => {
    return components.map((item) => {
      let loopType = type;
      let loopId = parentLoopId;
      if (LOOPCOMPONENTS.includes(item.type)) {
        loopType = !type ? 'outerLayer' : 'innerLayer';
        loopId = item.id!;
      }
      return {
        ...item,
        [LoopMarkSymbol]: {
          parentLoopId,
          loopType,
        },
        components: marker(item.components || [], loopId, loopType),
      };
    });
  };

  schema.components = marker(schema.components || [], '');
};

// 解析schema数据
export const parseSchema = (schema: IProjectSchema, isRoot: boolean) => {
  modifySchemaCompName(schema, isRoot);

  const target = preprocessComponentSchema(schema, isRoot) ?? schema;
  const { components } = target as any;
  target.components = components?.map((schem: any) => {
    return parseSchema(schem, false) ?? schem;
  });
  return target;
};

/**
 * 修改schema的组件名称
 * @param schema
 * @param isRoot
 */
const modifySchemaCompName = (schema: IProjectSchema, isRoot: boolean) => {
  const compName = schema?.compName || schema?.type;
  // path 在数据清理中已删除
  // if (schema.path) {
  //   // @ts-ignore
  //   delete schema.path;
  // }
  if (!isRoot && compName) {
    if (schema.compName) {
      schema.compName = compName;
    }
    schema.type = compName;
    if (schema?.customClass) {
      schema.props.className = schema.id;
    }
  }
};
