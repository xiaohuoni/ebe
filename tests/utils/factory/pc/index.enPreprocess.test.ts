import pc from '../../../../src/utils/factory/pc/index.enPreprocess';
import { extraData } from '../../../../src/utils/schema/lxschema';

describe('pc enPreprocess', () => {
  test('compilation completed', () => {
    const compDSL = {
      id: 'test-id',
      compName: 'Table',
      props: {
        rowKey: 'data.id',
        columns: [
          {
            attr: {
              attrNbr: 'data.attr1',
              childKeys: ['data.attr2'],
            },
          },
          {
            options: ['option1', 'option2'],
          },
        ],
        rowSelection: true,
        page: true,
        pageSize: 10,
        style: {
          color: 'red',
        },
      },
      style: {
        fontSize: '16px',
      },
    };

    const result = pc.Table(compDSL, extraData);

    expect(result).toMatchSnapshot();
  });
});
