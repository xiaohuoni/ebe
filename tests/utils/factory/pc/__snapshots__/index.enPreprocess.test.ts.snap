// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pc enPreprocess compilation completed 1`] = `
Object {
  "compName": "Table",
  "id": "test-id",
  "props": Object {
    "columns": Array [
      Object {
        "attr": Object {
          "attrNbr": "data.attr1",
          "childKeys": Array [
            "data.attr2",
          ],
        },
      },
      Object {
        "options": Array [
          "option1",
          "option2",
        ],
      },
    ],
    "page": true,
    "pageSize": 10,
    "rowKey": "data.id",
    "rowSelection": true,
    "style": Object {
      "color": "red",
    },
  },
  "style": Object {
    "fontSize": "16px",
  },
}
`;
